R/schema_autobox.R

```diff
@@ -0,0 +1,216 @@
+  if (!inherits(config, "config")) {
+    cli::cli_abort(
+      c("x" = "{.arg config} must be an object of class {.cls config}
+             not {.cls {class(config)}}.")
+    )
+  }
+  schema_id <- attr(config, "schema_id")
+  schema <- download_tasks_schema(
+    hubUtils::extract_schema_version(schema_id)
+  )
+
+  # Get list of paths to config properties that can be arrays and may require
+  # boxing.
+  paths <- get_array_schema_paths(schema)
+  if (!is.null(box_extra_paths)) {
+    paths <- c(paths, box_extra_paths) |> unique()
+  }
+
+  # Identify and append any additional properties that may need to be boxed.
+  paths <- append_additional_properties(config, paths)
+
+  paths <- paths |>
+    purrr::map(~ expand_path_items(.x, config)) |>
+    purrr::list_flatten()
+
+  # For every boxable path that exists in the config and has length 1L, coerce
+  # to list. This will force such elements to be written out as an array in JSON
+  # files.
+  for (path in paths) {
+    if (purrr::pluck_exists(config, !!!path)) {
+      config <- purrr::modify_in(
+        .x = config,
+        .where = path,
+        .f = box
+      )
+    }
+  }
+  return(config)
+}
+
+#' Get potential paths to properties in a config file that can be arrays from a hubverse schema
+#'
+#' The functions identifies properties in a hubverse schema that can be arrays of
+#' vectors and returns the paths of such elements in a config. Properties that can
+#' be arrays of objects are ignored.
+#' Useful to determine elements in a config object that may need to be boxed.
+#' Note that any `"items"` elements indicate that the property is an array of objects
+#'  and will be expanded with element index when applied to config objects.
+#' @param schema a list representation of a hubverse schema
+#'
+#' @return a list where each element is character vector of a path to a property
+#'  in the schema that can be an array of vectors. Elements returned as `"items"`
+#' @export
+#' @examplesIf curl::has_internet()
+#' schema <- download_tasks_schema("v3.0.1")
+#' get_array_schema_paths(schema)
+get_array_schema_paths <- function(schema) {
+  checkmate::assert_list(schema)
+  is_array_recursive(schema) |>
+    unlist() |>
+    names() |>
+    stringr::str_replace_all("properties\\.|oneOf\\.", "") |>
+    unique() |>
+    stringr::str_split("\\.")
+}
+
+# Check that a list element can be an array. Only returns TRUE when the element
+# can be an array of values (i.e. a leaf), not an array of objects (i.e. a node).
+can_be_array <- function(x) {
+  "array" %in% x[["type"]] && !"object" %in% x[["items"]][["type"]]
+}
+
+is_array_recursive <- function(x) {
+  if (is.atomic(x)) {
+    return(NULL)
+  }
+  if (can_be_array(x)) {
+    x <- TRUE
+  }
+  if (inherits(x, "list")) {
+    x[] <- purrr::map(x, is_array_recursive)
+  }
+  x
+}
+
+box <- function(x) {
+  if (is.list(x)) {
+    return(x)
+  }
+  if (is.null(x)) {
+    return(x)
+  }
+  if (length(x) > 1L) {
+    return(x)
+  }
```

Comment:
could this statement be simplified?

```suggestion
  if (is.list(x) || is.null(x) || length(x) > 1L) {
    return(x)
  }
```